### DISCUSSION TITLE
Template block visibility and conditional rendering

### DISCUSSION THREAD

**Discussion by joshuatf** [woocommerce/woocommerce#39855] [WooCommerce New Product Editor] [answered]
With the introduction of our new templating API, third party developers can interact with admin templates by adding new blocks, updating existing ones, or removing blocks from the template.

However, this does not address changing block visibility during runtime.  For example, a plugin may add a toggle to allow donations, which in turn hides the pricing fields.  This is a change that occurs in the client after the user has toggled the field.

## Template utils

As a https://github.com/woocommerce/woocommerce/pull/39517, I put together a set of client-side template utilities that can be utilized by templates and developers who want to interact with them.

This currently exposes a few methods for use with the template provider:

* `hideBlock( id )` - Hides a template block by a given ID.
* `unhideBlock( id )` - Unhides a template block given a block ID.
* `selectTemplate( templateId )` - Swaps the template out to with another template by ID.

## Client vs server-side conditional rendering

We've discussed a couple ways of handling this; one is a much more server-oriented approach while the other transfers a good chunk of the logic to the client-side.

Both approaches will likely have similar APIs as far as the server-side templating goes.

```php
$block = $template->get_block( 'block-to-hide-id' )->hide_if(
	[
		[
			'subject'  => 'product.quantity',
			'operator' => '>',
			'operand'  => 5,
		],
	]
);
```

### Server-side approach

Using the server-side approach, when a condition is applied to a block, we simply wrap that block in a [conditional block](https://github.com/woocommerce/woocommerce/blob/10ccbcce81be4e938ba0352e24a17f856711c97d/packages/js/product-editor/src/blocks/conditional/block.json).

```php
[
	'blockName'   => 'woocommerce/conditional',
	'args'        => [
		conditions' => [
			[
				'subject'  => 'product.quantity',
				'operator' => '>',
				'operand'  => 5,
			],
		],
	],
	'innerBlocks' => [
		[
			'blockName'   => 'block-to-hide-id',
		]
	]
]
```

This conditional block already exists and is being used to toggle certain fields on or off when a product has toggled the schedule sale price option.

**Advantages**

* Simple template-based approach that does not require further processing in the client to show or hide blocks.
* Debugging this is far easier since we can see conditions within the blocks and even visualize the conditional blocks if necessary to help debug.

**Disadvantages**

* Somewhat limited in the conditions that determine block visibility.  The subject used must be passed via context and the operand is static after it leaves the server.
* Wrapping in conditional blocks more deeply nests the template tree.  While this is mostly alleviated with the templating API since we never deal directly with the template as an array, it has the potential to make debugging the template structure more challenging should there be any bugs in the templating API.
### Client-side approach

The client-side approach essentially is just dog-fooding the template utils mentioned above.

```js
// Template utils.
function evaluateCondition( condition, context ) {
	// Return true or false depending on condition.
};
function evaluateConditions( blockConditions, context ) {
	const { blockId, conditions } = blockConditions;
	const shouldHide = conditions.reduce( ( result, condition ) => {
		if ( evaluateCondition( condition ) ) {
			return true;
		}
		return result;
	}, false );
	if ( shouldHide ) {
		hideBlock( blockId );
	} else {
		unhideBlock( blockId );
	}
}

// Template provider.
useEffect( () => {
	const { blockConditions } = template;
	evaluateConditions( blockConditions, context );
}, [ context ] );

// Third party blocks can also implement their own custom logic for hiding/unhiding blocks.
function Edit() {
	useEffect( () => {
		if ( someCondition ) {
			hideBlock( 'block-to-hide' );
			return;
		}
		unhideBlock( 'block-to-hide' );
	}, [ someCondition ] );
}
```

**Advantages**

* Slightly more performant as it does not require blocks parsing additional conditional blocks and their children.
* Simplifies the backend templating logic and removes an additional methods needed to handle conditional wrapping of inner blocks.
* Allows more control around unhiding blocks that have been hidden by other block conditions.

**Disadvantages**

* Does not provide the option to show or hide blocks visually.  I.e., hiding a block completely removes it from the render, whereas the conditional block can hide the visibility of the block, but allow its side effects to still run.
* Requires an additional field in the template REST API response to house conditions. 

## Handling condition conflicts

One issue we have with both approaches is the conflict between registered conditions and manual toggling block visibility.  For example, imagine that we have a block with `some-block` as an ID with the following condition to hide the block:

```php
[
	'subject'  => 'product.quantity',
	'operator' => '>',
	'operand'  => 5,
]
```

Then we set the product quantity to 10, while another plugin has added some logic within their block:

```js
function Edit() {
	useEffect( () => {
		if ( someCondition ) {
			hideBlock( 'some-block' );
			return;
		}
		unhideBlock( 'some-block' );
	}, [ someCondition ] );
}
```

When `someCondition` is `false` we will unhide the block, regardless of the other conditions that were previously registered.

On a similar note, when we use hide or unhide functions on a block that has been hidden by the server and is wrapped with a conditional block, those functions will not work since they are acting on the block itself and not the conditional wrapping block.

Which of these makes more sense?  **Unhiding a block when another condition says it should remain hidden or the unhiding method not working?**

IMO, the latter makes slightly more sense, but both cases are a little odd.  Despite this, I still think the client-side conditions might be a better approach since the hiding/unhiding works directly on the block in question as opposed to the parent block.  This also leaves room to handle conflict resolution client-side (and implement custom behavior here) whereas the conditional block complicates this API significantly by needing to make changes to the parent block.

## Thoughts

I don't feel too strongly on which approach we use as they seem to both accomplish our goals, however I think it is clear that even with the server-side approach we still need the client-side utils so that we have finer control over templates in the client.

With that said, I'm slightly leaning towards the client-side conditions.  The extra wrapping conditional block seems like it opens room for a lot of headaches and added layer of complexity in dealing with the templates.

**Comment by mattsherman** [comment]
So, in both cases, we'd have the ability to specify the conditional in the template server-side?

We're talking about how the visibility would be implemented?

As far as handling conflicts... what if the visibility of a block was a filter that other client code could hook into? Then it could be handled via priorities as is done with filters.

> Does not provide the option to show or hide blocks visually. I.e., hiding a block completely removes it from the render, whereas the conditional block can hide the visibility of the block, but allow its side effects to still run.

Can you think of any examples where we'd want a block to be visually hidden but still have its side effects run?

**Reply by octaedro** [reply]
> what if the visibility of a block was a filter that other client code could hook into? Then it could be handled via priorities as is done with filters.

I like the idea of providing a comprehensive list of conditions that determine whether to hide or unhide a block.

**Reply by joshuatf** [reply]
>So, in both cases, we'd have the ability to specify the conditional in the template server-side?

Yep!  I think this is necessary so that various plugins can interact with visibility conditions.  Which in turn makes me wonder if we should even allow functionality like `hideBlock` without some preset condition due to potential conflicts.  But I also think those methods could be potentially useful despite (or especially due to) overriding conditions 🤔 

>As far as handling conflicts... what if the visibility of a block was a filter that other client code could hook into? Then it could be handled via priorities as is done with filters.

I really would prefer this route.  But AFAIK filters are still unpredictable in the frontend due to the asynchronous nature of how they're loaded during runtime.  I was under the impression we were supposed to be moving away from them, but maybe this is up for debate.

>Can you think of any examples where we'd want a block to be visually hidden but still have its side effects run?

Somewhat contrived scenario, but I'm sure there are others: The product name block sets the SKU if the name has been changed.  If this block was hidden because a third party name field was added, we'd probably still want this logic to run.

But I think there's also a strong argument for moving this logic elsewhere and allowing this logic to be filled or filtered.  That's something I'd like to see long term.

**Reply by joshuatf** [reply]
>I like the idea of providing a comprehensive list of conditions that determine whether to hide or unhide a block.

@octaedro Do you think this means we should limit the ability of methods like `hideBlock` or `unhideBlock` and defer to set conditions only?

**Reply by mattsherman** [reply]
> Somewhat contrived scenario, but I'm sure there are others: The product name block sets the SKU if the name has been changed. If this block was hidden because a third party name field was added, we'd probably still want this logic to run.

> But I think there's also a strong argument for moving this logic elsewhere and allowing this logic to be filled or filtered. That's something I'd like to see long term.

Agreed. Ideally, for the example where the sku is set when the name has been changed, I don't think that logic should be in the name block. Is there any existing concept in Gutenberg where such logic could reside? Could we hook into the entity store somehow to do such a thing?

I guess the main issue with moving the logic elsewhere would be making it easy for devs to know how to enqueue this logic.

**Comment by octaedro** [comment]
Thank you @joshuatf, for summarizing this.

Considering the restrictions we have, at first glance, I lean towards the second option, which is client-side rendering, but I would love to see a server-side rendering solution.

> Which of these makes more sense? Unhiding a block when another condition says it should remain hidden or the unhiding method not working?

If we don't have any other option, I would prefer the latter. However, I'd be interested in exploring the concept that @mattsherman mentioned above:  a hook (or something) to show 3PD all the rules applied to a block.

**Reply by joshuatf** [reply]
>However, I'd be interested in exploring the concept that @mattsherman mentioned above: a hook (or something) to show 3PD all the rules applied to a block.

I am too.  Though any third party code that adds a filter would likely need to be loaded and ready prior to the templates being fetched, otherwise the filter would not be run correctly.

One way around this is to use something like Gutenberg's use of `__experimentalReapplyBlockTypeFilters` to make sure filters get reapplied at certain events.

**Comment by louwie17** [comment]
Thanks for starting this discussion here @joshuatf. 

In your client side approach example you make use of `onChange( visibleBlocks, {} );`, which updates the entire template structure. Do we know what the performance impact of this is? It may not be used, as it would be identical to users adding/removing a paragraph in the full site editor ( but just curious ).
I am only pointing this out, as this approach actually changes the template tree, while the other (wrapping conditional blocks) doesn't and keeps the context of showing/hiding local to the conditional block.

I suppose it goes back to your point of where `hiding a block completely removes it from the render`, and if there is a strong case to be made that we don't want that to happen.
This kinda makes me think that conditional logic should not be part of the block render it-self, and we shouldn't expose it there either.
Instead we could provide a utility outside of the template render that handles this, so we never have to worry about certain conditionals not working anymore because the block was removed.
I am not a 100% sure how this would look, but may make it easier to follow what @mattsherman suggested, by using something similar as a filter approach.
From second glance at your client side example, it seems you were already doing something similar, where the conditionals were handled outside of the block context.

On the PHP end we have the same API for registering conditionals, upon parsing of the template we can filter these conditionals out and expose them in their own list, with a potential priority added.
There would still be a case to add additional conditional client side, although those should not be registered within the `Edit` component, but we can have something similar as `registerWooBlockType` like `registerWooBlockConditional`.
This can listen to any changes on the product and call `onChanges` to change the template structure.

The only sucky thing is that an approach like this would still have to be run `client` side, if we want to give the `registerWooBlockConditional` the same priority as the PHP registered conditionals. Although I suppose we can still have an additional PHP filter similar to `registerWooBlockConditional`, that is mostly for custom cases, so things can still be parsed on the server side.

**Reply by joshuatf** [reply]
>Do we know what the performance impact of this is? It may not be used, as it would be identical to users adding/removing a paragraph in the full site editor ( but just curious ).

I'm not sure and I haven't seen a lot of tests on the Gutenberg side, so this may be something we need to test ourselves to determine if it has meaningful impact.

I do think that usage of block visibility utils is akin to switching an entire template since many of these use cases would involve the product "traits" that toggle on/off various elements in the form, so the same performance impact of switching may be acceptable for those cases.

But thinking about this further and taking some concepts from the [universal block registration](https://github.com/woocommerce/woocommerce/discussions/39775), I think we can optimize this a good amount with a wrapper around the blocks:

```js
registerWooBlockType( block ) {
    const { edit } = block;

    registerBlockType( {
        ...block,
        edit: WooEditWrapper( edit ),
    } );
}
function WooEditWrapper( edit ) {
    return ( props ) {
        const { _templateId } = props.attributes;
        if ( hiddenTemplateBlocks.includes( _templateId ) ) {
            return null;
        }
        return <edit { ...props } />
    }
}
```

The above is a simple example, but could be extended to incorporate `conditionals` in the wrapper.

>I suppose it goes back to your point of where hiding a block completely removes it from the render, and if there is a strong case to be made that we don't want that to happen.

I'm now leaning pretty strongly towards removing it completely from render.  After discussing with @mattsherman above, I think cases where we want side effects to run despite the block not being rendered represent logic that should be moved outside of the block.  Not worrying about "rendering but hiding" simplifies this process a bit and improves performance.

>The only sucky thing is that an approach like this would still have to be run client side, if we want to give the registerWooBlockConditional the same priority as the PHP registered conditionals. 

Ya, it's starting to feel like we're creating 2 versions of the same API, one for the client and one for the server.  Priority becomes an issue as does ongoing maintenance for 2 separate APIs.

A filter does simplify this process a good amount.  I'll look around and see if this is appropriate for us to use.

**Reply by louwie17** [reply]
> I'm now leaning pretty strongly towards removing it completely from render. After discussing with @mattsherman above, I think cases where we want side effects to run despite the block not being rendered represent logic that should be moved outside of the block. Not worrying about "rendering but hiding" simplifies this process a bit and improves performance.

Makes sense, I think it would be worth sticking with this approach for now. Given users can always manually do the other approach for their own blocks if they really wanted to, by handling the show/hide within the render method.

**Comment by louwie17** [comment]
> Does not provide the option to show or hide blocks visually. I.e., hiding a block completely removes it from the render, whereas the conditional block can hide the visibility of the block, but allow its side effects to still run.

One additional suggestion to circumvent the block removal completely. Each block has a `class` that is derived from the block name, so if we wanted, we could just add styling on the fly and target these specific classes with `display: none`.
The only slight problem is that it doesn't solve the use cases of multiple of the same blocks that use different conditions. That could be fixed if we can easily add the unique `id` as a class as well.
Although I suppose we could also do that by just adding the `id` as `className` to `attributes`: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/use-block-props/use-block-custom-class-name.js#L19

**Comment by joshuatf** [comment, marked as answer]
@mattsherman is taking over this during my AFK.  The current plan is as follows:

* We will allow methods on the templates in PHP that accept conditional hiding of blocks.  E.g., `$block->hide_if( ... )`
* The conditions will be passed to the client, allowing any context properties to be evaluated by the registered condition.  Prior art for using block context and an evaluator can be found in https://github.com/woocommerce/woocommerce/pull/39461
* The actual hiding of blocks will be handled via a wrapper around the `Edit` component mentioned in https://github.com/woocommerce/woocommerce/discussions/39855#discussioncomment-6815835 (I will have a POC for contextual blocks that demonstrates this posted later today)

No client-side specific methods or filters for hiding/unhiding blocks will be added at the moment.  The reason for this choice is that adding methods in the client would override any visibility logic added to the template via PHP and currently we don't have a use case for this.  If we decide to pursue this further, a client-side filter is most likely the best choice, but we will need to carefully integrate the PHP added conditions within this filter and determine a default priority for those.

**Reply by joshuatf** [reply]
Here is the POC for contextual blocks that shows a wrapper for `Edit` components capable of handling visibility.  https://github.com/woocommerce/woocommerce/pull/40077
